File: server/src/routes/influencers.ts
import { Router, Request, Response, NextFunction } from 'express';
import { eq, desc, sql } from 'drizzle-orm';
import db from '../db/index.js';
import { influencers } from '../db/schema.js';
import { authenticate, authorize } from '../middleware/auth.js';
import { AppError } from '../middleware/error.js';

const router = Router();

// Search influencers (with filters, pagination, sorting)
router.get('/', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { category, niche, minFollowers, maxRate, search, page = '1', limit = '20', sort = 'totalFollowers' } = req.query;
    const offset = (parseInt(page as string) - 1) * parseInt(limit as string);

    let results = await db.select().from(influencers).orderBy(desc(influencers.totalFollowers)).limit(parseInt(limit as string)).offset(offset);

    if (search) results = results.filter(i => i.displayName.toLowerCase().includes((search as string).toLowerCase()) || i.category?.toLowerCase().includes((search as string).toLowerCase()));
    if (category && category !== 'All') results = results.filter(i => i.category === category);
    if (niche) results = results.filter(i => i.niche === niche);
    if (minFollowers) results = results.filter(i => i.totalFollowers >= parseInt(minFollowers as string));
    if (maxRate) results = results.filter(i => parseFloat(String(i.pricePerPost || 0)) <= parseInt(maxRate as string));

    const countResult = await db.select({ count: sql<number>`count(*)` }).from(influencers);

    res.json({
      success: true,
      data: results,
      pagination: { page: parseInt(page as string), limit: parseInt(limit as string), total: countResult[0]?.count || 0 },
    });
  } catch (err) { next(err); }
});

// Get influencer by ID
router.get('/:id', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const result = await db.select().from(influencers).where(eq(influencers.id, req.params.id)).limit(1);
    if (!result.length) throw new AppError('Influencer not found', 404);
    res.json({ success: true, data: result[0] });
  } catch (err) { next(err); }
});

// Update influencer profile
router.patch('/profile', authenticate, authorize('influencer'), async (req: Request, res: Response, next: NextFunction) => {
  try {
    const inf = await db.select().from(influencers).where(eq(influencers.userId, req.user!.sub)).limit(1);
    if (!inf.length) throw new AppError('Profile not found', 404);

    const { displayName, bio, category, niche, location, totalFollowers, engagementRate, pricePerPost, audienceAgeGroup, audienceGender, socialLinks, platforms } = req.body;

    await db.update(influencers).set({
      ...(displayName && { displayName }),
      ...(bio !== undefined && { bio }),
      ...(category && { category }),
      ...(niche && { niche }),
      ...(location && { location }),
      ...(totalFollowers !== undefined && { totalFollowers }),
      ...(engagementRate !== undefined && { engagementRate }),
      ...(pricePerPost !== undefined && { pricePerPost }),
      ...(audienceAgeGroup && { audienceAgeGroup }),
      ...(audienceGender && { audienceGender }),
      ...(socialLinks && { socialLinks }),
      ...(platforms && { platforms }),
    }).where(eq(influencers.id, inf[0].id));

    const updated = await db.select().from(influencers).where(eq(influencers.id, inf[0].id)).limit(1);
    res.json({ success: true, data: updated[0] });
  } catch (err) { next(err); }
});

export default router;

File: Backend/routes/influencers.py
import os
import uuid
from flask import Blueprint, request, jsonify, current_app
from flask_jwt_extended import jwt_required, get_jwt_identity, get_jwt
from werkzeug.utils import secure_filename
from models import db
from models.influencer import Influencer

influencers_bp = Blueprint('influencers', __name__)

ALLOWED = {'png', 'jpg', 'jpeg'}

def allowed_file(filename):
    return '.' in filename and filename.rsplit('.', 1)[1].lower() in ALLOWED


@influencers_bp.route('', methods=['GET'])
def search_influencers():
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('limit', 20, type=int)
    category = request.args.get('category')
    location = request.args.get('location')
    min_followers = request.args.get('minFollowers', type=int)
    max_rate = request.args.get('maxRate', type=float)
    search = request.args.get('search')
    sort_by = request.args.get('sort', 'followers')

    query = Influencer.query
    if category and category != 'All':
        query = query.filter(Influencer.category == category)
    if location:
        query = query.filter(Influencer.location.ilike(f'%{location}%'))
    if min_followers:
        query = query.filter(Influencer.total_followers >= min_followers)
    if max_rate:
        query = query.filter(Influencer.price_per_post <= max_rate)
    if search:
        query = query.filter(
            db.or_(
                Influencer.display_name.ilike(f'%{search}%'),
                Influencer.category.ilike(f'%{search}%'),
                Influencer.niche.ilike(f'%{search}%'),
            )
        )

    if sort_by == 'engagement':
        query = query.order_by(Influencer.engagement_rate.desc())
    elif sort_by == 'score':
        query = query.order_by(Influencer.previous_campaign_score.desc())
    else:
        query = query.order_by(Influencer.total_followers.desc())

    pagination = query.paginate(page=page, per_page=per_page, error_out=False)

    return jsonify(
        success=True,
        data=[i.to_dict() for i in pagination.items],
        pagination={'page': page, 'limit': per_page, 'total': pagination.total, 'pages': pagination.pages}
    ), 200


@influencers_bp.route('/<influencer_id>', methods=['GET'])
def get_influencer(influencer_id):
    influencer = Influencer.query.get(influencer_id)
    if not influencer:
        return jsonify(success=False, error='Influencer not found'), 404
    return jsonify(success=True, data=influencer.to_dict()), 200


@influencers_bp.route('/profile', methods=['PUT'])
@jwt_required()
def update_profile():
    claims = get_jwt()
    if claims.get('role') != 'influencer':
        return jsonify(success=False, error='Influencer role required'), 403

    user_id = get_jwt_identity()
    influencer = Influencer.query.filter_by(user_id=user_id).first()
    if not influencer:
        return jsonify(success=False, error='Profile not found'), 404

    data = request.get_json()
    for field, attr in [
        ('displayName', 'display_name'), ('bio', 'bio'), ('category', 'category'),
        ('niche', 'niche'), ('location', 'location'), ('audienceAgeGroup', 'audience_age_group'),
        ('audienceGender', 'audience_gender'),
    ]:
        if field in data:
            setattr(influencer, attr, data[field])

    if 'totalFollowers' in data:
        influencer.total_followers = int(data['totalFollowers'])
    if 'engagementRate' in data:
        influencer.engagement_rate = float(data['engagementRate'])
    if 'pricePerPost' in data:
        influencer.price_per_post = float(data['pricePerPost'])
    if 'socialLinks' in data:
        influencer.set_social_links(data['socialLinks'])

    db.session.commit()
    return jsonify(success=True, data=influencer.to_dict()), 200


@influencers_bp.route('/upload-image', methods=['POST'])
@jwt_required()
def upload_image():
    claims = get_jwt()
    if claims.get('role') != 'influencer':
        return jsonify(success=False, error='Influencer role required'), 403

    if 'file' not in request.files:
        return jsonify(success=False, error='No file provided'), 400

    file = request.files['file']
    if file.filename == '':
        return jsonify(success=False, error='No file selected'), 400

    if not allowed_file(file.filename):
        return jsonify(success=False, error='File type not allowed. Use jpg, jpeg, or png'), 400

    upload_dir = current_app.config['UPLOAD_FOLDER']
    os.makedirs(upload_dir, exist_ok=True)

    ext = file.filename.rsplit('.', 1)[1].lower()
    filename = f"{uuid.uuid4().hex}.{ext}"
    filepath = os.path.join(upload_dir, filename)
    file.save(filepath)

    image_url = f"/uploads/{filename}"

    user_id = get_jwt_identity()
    influencer = Influencer.query.filter_by(user_id=user_id).first()
    if influencer:
        influencer.profile_image = image_url
        db.session.commit()

    return jsonify(success=True, data={'url': image_url, 'filename': filename}), 200
